### DISCUSSION TITLE
Ease search results navigation by page selection

### DISCUSSION THREAD

**Discussion by curquiza** [meilisearch/product#483] [Feedback & Feature Proposal]
I open this discussion for all the users that are currently not satisfied with the Meilisearch pagination, I mean using the `limit` and `offset` parameters. Indeed, we noticed some users might need access to the last page of their data, something that they currently cannot do with Meilisearch. Only the "next" page is accessible using `offset` and `limit`

We also noticed some users use `nbHits` to access the end of the data, something we definitely NOT recommend, [see why here](https://github.com/meilisearch/product/discussions/467).

In v0.28.0 we decided to rename `nbHits` to `estimatedNbHits` to avoid the current confusion our users have with this parameter which should not be used for pagination. Check out [this discussion](https://github.com/meilisearch/product/discussions/394) for more details.

Knowing this, if you, as a website developer integrating Meilisearch into it, would need a kind of pagination Meilisearch cannot handle currently (or handle with a tricky workaround), please let us know here your use case and what you would expect. This would definitely be helpful for us ❤️

**Comment by eelcol** [comment]
I am using Meilisearch to display all products on the site, so not only in the search functionality. I use pagination for example on the category page. I want the user to be able to scroll to every available page. Although it is not recommended, I calculate the number of pages based on 'nbHits'. It works for me, but a 'real' solution for this would be nice.

**Reply by gmourier** [reply]
Thanks @eelcol ! 

"I want the user to be able to scroll to every available page." Can you tell us why you need to do that with a page selector component? (Aka letting users select a specific page while other methods still let them scroll to the end; infinite scrolling / load more in mind)

**Comment by pruhstal** [comment]
As similar with @eelcol, I show a default view for a couple of product overview pages that I would like by default to be paged and sorted alphabetically (before the user filters with their keywords).

**Reply by gmourier** [reply]
Thanks, @pruhstal. I have the same question as for @eelcol. 😇

Can you tell us why you need to do that with a page selector component? (It seems to me that you want to let users select a specific page)

**Reply by pruhstal** [reply]
Hi @gmourier! I'll try to best present my case with the following points:

1. Maintaining scroll position with an "infinite" list of items when navigating back to that list has always been historically hacky. Not to mention the more you scroll, the more event listener management needs to be handled (e.g. attaching scroll events to unbind events not in the viewport and rebinding when in viewport).
2. SEO becomes simpler. I don't have to list every single known URL in multiple sitemaps (depending on how much content I have). I can list a single page, and the crawler can discover the paginated links.
3. This is a more commerce-centric website. Like most major e-commerce websites (Amazon, Flipkart, Alibaba, eBay, etc), they are using traditional pagination instead of an infinite scroll/load more implementation to discover products. This is mostly for accessibility (a11y) purposes.
5. I have a footer that's not fixed nor I intend it to be. It "sticks" to the bottom of the content div so when you reach the end of the page, it can be discovered. With infinite scroll, it's not discoverable, and I am not a fan of a "load more" type hybrid content discovery which would be the only other way to currently build "pagination" with Meilisearch.

Happy to elaborate on this a bit more, but ultimately, pagination gives the user more control.

**Reply by gmourier** [reply]
Thanks, @pruhstal 🙏 That's really clear.

**Comment by ManyTheFish** [comment]
Hello @eelcol and @pruhstal!
We are working on improving pagination in Meilisearch and we just released a first prototype that makes creating numbered pagination way easier! :fire:
It would be super helpful if you could let us know if this helps with your issue.
To know more about the prototype and how to use it, take a look at [the dedicated discussion](https://github.com/meilisearch/meilisearch/discussions/2635).

**Comment by mech** [comment]
I am using this Meilisearch instead of Elasticsearch for quick wins on my Applicant Tracking System of several millions of candidates. I too would love to get a consistent traditional pagination solution.

**Reply by ManyTheFish** [reply]
Hello @mech!
We are working on improving pagination in Meilisearch and we just released a first prototype that makes creating numbered pagination way easier! :fire:
It would be super helpful if you could let us know if this helps with your issue.
To know more about the prototype and how to use it, take a look at [the dedicated discussion](https://github.com/meilisearch/meilisearch/discussions/2635).

**Comment by m4tthumphrey** [comment]
I am not sure if this is the right discussion to mention this but I came here from #467...

I am actually using Meilisearch in a slightly unorthodox way. My data consists of ~11m documents whose attributes are all integers. I use MS because of its speed and was hoping to use the `estimatedTotalHits` property as a very efficient way of performing statistic based searches, ie I do not always need each document in the response, just the count of documents.

**Reply by ManyTheFish** [reply]
Hello @m4tthumphrey in the current version of Meilisearch you can't have an exhaustive count of hits.
However, the next release of Meilisearch (v0.30) will allow you to compute this exhaustive count using page selection.
If you want to try this feature in beta, [we've created a prerelease today!](https://github.com/meilisearch/meilisearch/releases/tag/v0.30.0rc0)

> You can find a minimal documentation in the **Ease search results navigation with pagination** section.
Note that the first page is page 1, however, asking for `page` 0 will compute a search query and return the exhaustive value of `totalPages` and `totalHits` without returning any document

But if you don't need an exhaustive count, I'll suggest just setting a `limit` to 0 in query parameters, this would return the `estimatedTotalHits` without any documents.

**Reply by m4tthumphrey** [reply]
@ManyTheFish That limit = 0 tip is great, thank you!

Wow that release has loads of things that I am waiting for, especially the cancelling of tasks!

I look forward to trying it out, do you have any idea of when 1.30 stable will be released?

**Reply by curquiza** [reply]
You can already test the RC of v0.30.0 -> https://github.com/meilisearch/meilisearch/releases/tag/v0.30.0rc0

We do not recommend using it in production since it's a pre-release, but you can start development and let us know about any issues you might encounter.
The official v0.30.0 release will be done 28th November.

**Reply by mech** [reply]
Wonderful, I can test both filtering and pagination at the same time now 👍

**Comment by gmourier** [comment]
Hey everyone 👋

[`v0.30`](https://github.com/meilisearch/meilisearch/releases/tag/v0.30.1) introduced a new way to navigate search results. It's now possible to implement an exhaustive page selection component. 🪄 

📚 More information [here](https://docs.meilisearch.com/learn/advanced/pagination.html#numbered-page-selectors).

**Comment by gmourier** [comment]
Hello everyone 👋

I'm locking that discussion.

Please, open a new discussion if you have feedback to share.

Thank you 🙇‍♂️
